orm/genesis.go

```diff
@@ -29,85 +19,58 @@ type SequenceExportable interface {
 	Sequence() Sequence
 }
 
-// ExportTableData returns a json encoded `[]Model` slice of all the data persisted in the table.
+// ExportTableData iterates over the given table entries and stores them at the passed ModelSlicePtr.
 // When the given table implements the `SequenceExportable` interface then it's current value
 // is returned as well or otherwise defaults to 0.
-func ExportTableData(ctx HasKVStore, t TableExportable) (json.RawMessage, uint64, error) {
-	enc := jsonpb.Marshaler{}
-	var r []Model
-	err := forEachInTable(ctx, t.Table(), func(rowID RowID, obj codec.ProtoMarshaler) error {
-		var buf bytes.Buffer
-		err := enc.Marshal(&buf, obj)
-		if err != nil {
-			return errors.Wrap(err, "json encoding")
-		}
-		r = append(r, Model{Key: rowID, Value: buf.Bytes()})
-		return nil
-	})
+func ExportTableData(ctx HasKVStore, t TableExportable, dest ModelSlicePtr) (uint64, error) {
+	it, err := t.Table().PrefixScan(ctx, nil, nil)
 	if err != nil {
-		return nil, 0, err
+		return 0, errors.Wrap(err, "all rows prefix scan")
+	}
+	_, err = ReadAll(it, dest)
+	if err != nil {
+		return 0, err
 	}
-
 	var seqValue uint64
 	if st, ok := t.(SequenceExportable); ok {
 		seqValue = st.Sequence().CurVal(ctx)
 	}
-	b, err := json.Marshal(r)
-	return b, seqValue, err
+	return seqValue, err
 }
 
-// ImportTableData initializes a table and attached indexers from the given json encoded `[]Model`s.
+// ImportTableData initializes a table and attaches indexers from the given data interface{}.
```

Comment:
> * use `[]NaturalKeyed`  (that would require copying slice all whenever we call this function - which shouldn't be a big deal, because we will only copy `len(slice)` pointers).

As you said, for example that would mean that we would need to copy the slice of `GroupInfo` into a slice of `Model` before calling the `ImportTableData` function, and I don't find this really good from a dev UX point of view. Also it would just clutter the `InitGenesis` function in `x/group/server/genesis.go`, because we would need to add that for all the table imports.

> * or an interator pattern:
> 
> ```
> type NaturalKeyedIterator interface {
>    Next() bool   // goes to the next element, returns false if there are no more elements
>    Get() NaturalKeyed // returns current element
> }
> 
> type MyObjs {
>    slice []MyObj
>    idx uint
> }
> func (m *MyObjs) Next () { 
>     if m.idx >= len(slice) - 1 { return false }
>     m.idx++
>     return true
> } 
> func (m *MyObjs) Get() NaturalKeyed {  return m.slice[m.idx] }
> ```

I prefer this second approach, but that would mean that we would need to define a struct such as `MyObjs` for all models, eg Groups, GroupAccounts, etc. while their `NaturalKeyedIterator` implementation logic would all be the same.

Reply:
I think copying indexes (first approach) is not bad. It won't take much more memory since we are only copying pointers.

If you like we can do it (or consider 2nd approach or even something else) later in a new issue?

Reply:
Ok I'll create an issue

Reply:
https://github.com/regen-network/regen-ledger/issues/308

Reply:
Oh, I did it in the meantime. https://github.com/regen-network/regen-ledger/issues/309

Reply:
ok I'll close mine